File: src/main/java/dev/kensa/util/KensaMap.java
package dev.kensa.util;

import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Stream;

import static dev.kensa.util.Attributes.emptyAttributes;

@SuppressWarnings({"unchecked", "WeakerAccess"})
public class KensaMap<M extends KensaMap<M>> {

    private static final Pattern KEY_PATTERN = Pattern.compile("(?<prefix>.*)(?<key>__(?<prekey>[ ]*)idx(?<postkey>[ ]*)__)(?<suffix>.*)");

    private final Object lock = new Object();
    private final Map<String, Entry> values = new LinkedHashMap<>();

    public M put(Object value) {
        Objects.requireNonNull(value);

        return putWithUniqueKey(value.getClass().getSimpleName() + "__idx__", value, emptyAttributes());
    }

    public M put(String key, Object value) {
        return this.put(key, value, emptyAttributes());
    }

    public M put(String key, Object value, Attributes attributes) {
        Objects.requireNonNull(key);
        Objects.requireNonNull(value);
        Objects.requireNonNull(attributes);

        synchronized (lock) {
            values.put(key, new Entry(key, value, attributes));
        }

        return self();
    }

    public M putWithUniqueKey(String key, Object value, Attributes attributes) {
        Objects.requireNonNull(value);

        Matcher matcher = KEY_PATTERN.matcher(key);
        if (matcher.matches()) {
            synchronized (lock) {
                String realKey = baseKeyFrom(matcher);
                if (values.containsKey(realKey)) {
                    int i = 1;
                    do {
                        realKey = indexedKeyFrom(matcher, i++);
                    } while (values.containsKey(realKey));
                }

                values.put(realKey, new Entry(realKey, value, attributes));

                return self();
            }
        } else {
            throw new IllegalArgumentException("Must specify __key__ placeholder");
        }
    }

    public <T> void putAll(Collection<T> values) {
        Objects.requireNonNull(values);

        synchronized (lock) {
            values.forEach(this::put);
        }
    }

    public void putNamedValues(Collection<NamedValue> values) {
        Objects.requireNonNull(values);

        synchronized (lock) {
            values.forEach(nv -> this.put(nv.name(), nv.value()));
        }
    }

    public <T> T get(String key, Class<T> clazz) {
        Entry entry = values.get(key);
        if (entry == null) {
            return null;
        }
        if (!clazz.isAssignableFrom(entry.value.getClass())) {
            throw new ClassCastException(String.format("Cannot cast actual type [%s] to requested type [%s]", entry.getClass().getName(), clazz.getName()));
        }

        return (T) entry.value;
    }

    public boolean containsKey(String key) {
        return values.containsKey(key);
    }

    public Set<Entry> entrySet() {
        return new LinkedHashSet<>(values.values());
    }

    private M self() {
        return (M) this;
    }

    private String baseKeyFrom(Matcher matcher) {
        return matcher.group("prefix") + matcher.group("suffix");
    }

    private String indexedKeyFrom(Matcher matcher, int index) {
        return matcher.group("prefix") + (matcher.group("prekey") + index +
                (Strings.isNotBlank(matcher.group("suffix")) ? matcher.group("postkey") : ""))
                + matcher.group("suffix");
    }

    public static class Entry {

        private final String key;
        private final Object value;
        private final Attributes attributes;

        public Entry(String key, Object value, Attributes attributes) {
            this.key = key;
            this.value = value;
            this.attributes = attributes;
        }

        public Entry(String key, Object value) {
            this(key, value, emptyAttributes());
        }

        public String key() {
            return key;
        }

        public Object value() {
            return value;
        }

        public Stream<NamedValue> attributes() {
            return attributes.stream();
        }
    }
}

File: src/main/java/dev/kensa/util/KensaMap.kt
package dev.kensa.util

import dev.kensa.util.Attributes.Companion.emptyAttributes
import java.util.*

abstract class KensaMap<M : KensaMap<M>> {

    private val lock = Any()
    private val values: MutableMap<String, Entry> = LinkedHashMap()

    fun put(value: Any, timestamp: Long = System.currentTimeMillis()): M = putWithUniqueKey(value.javaClass.simpleName + "__idx__", value,timestamp, emptyAttributes())

    @JvmOverloads
    fun put(key: String, value: Any?, timestamp: Long = System.currentTimeMillis(), attributes: Attributes = emptyAttributes()): M = self().apply {
        synchronized(lock) { values[key] = Entry(key, value, timestamp, attributes) }
    }

    internal fun putWithUniqueKey(key: String, value: Any?, timestamp: Long = System.currentTimeMillis(), attributes: Attributes = emptyAttributes()): M = self().apply {
        fun baseKeyFrom(result: MatchResult) = result.groups["prefix"]?.value + result.groups["suffix"]?.value
        fun indexedKeyFrom(result: MatchResult, index: Int) = (result.groups["prefix"]?.value + (result.groups["prekey"]?.value + index +
                if (result.groups["suffix"]?.value?.isNotBlank() == true) result.groups["postkey"]?.value else "") + result.groups["suffix"]?.value)

        KEY_REGEX.matchEntire(key)?.let { result ->
            synchronized(lock) {
                var realKey = baseKeyFrom(result)
                if (values.containsKey(realKey)) {
                    var i = 1
                    do {
                        realKey = indexedKeyFrom(result, i++)
                    } while (values.containsKey(realKey))
                }
                values[realKey] = Entry(realKey, value, timestamp, attributes)
            }
        } ?: throw IllegalArgumentException("Must specify __key__ placeholder")
    }

    fun putAll(values: Collection<Any>) {
        synchronized(lock) { values.forEach { value -> this.put(value) } }
    }

    fun putNamedValues(values: Collection<NamedValue>) {
        synchronized(lock) { values.forEach { nv: NamedValue -> this.put(nv.name, nv.value) } }
    }

    operator fun <T> get(key: String): T? =
            synchronized(lock) {
                values[key]?.let { entry ->
                    @Suppress("UNCHECKED_CAST")
                    entry.value as T
                }
            }

    fun containsKey(key: String): Boolean = synchronized(lock) { values.containsKey(key) }

    fun entrySet(): Set<Entry> = synchronized(lock) { LinkedHashSet(values.values.sortedBy(Entry::timestamp)) }

    @Suppress("UNCHECKED_CAST")
    private fun self(): M = this as M

    class Entry(val key: String, val value: Any?, val timestamp: Long, val attributes: Attributes = emptyAttributes())

    companion object {
        private val KEY_REGEX = "(?<prefix>.*)(?<key>__(?<prekey>[ ]*)idx(?<postkey>[ ]*)__)(?<suffix>.*)".toRegex()
    }
}